crates/rpc/rpc/src/eth/logs_utils.rs

```diff
@@ -1,46 +1,95 @@
+        let logs = &receipt.logs;
+        for log in logs {
+            if log_matches_filter(block_num_hash, log, filter) {
+                let log = Log {
+                    address: log.address,
+                    topics: log.topics.clone(),
+                    data: log.data.clone(),
+                    block_hash: Some(block_num_hash.hash),
+                    block_number: Some(U256::from(block_num_hash.number)),
+                    transaction_hash: Some(tx_hash),
+                    // The transaction and receipt index is always the same.
+                    transaction_index: Some(U256::from(receipt_idx)),
+                    log_index: Some(U256::from(log_index)),
+                    removed,
+                };
+                all_logs.push(log);
+            }
+            log_index += 1;
+        }
+    }
     all_logs
 }
 
-/// Appends all matching logs of a block's receipts grouped with the hash of their transaction
-pub(crate) fn append_matching_block_logs<'a, I>(
+/// Appends all matching logs of a block's receipts.
+/// If the log matches, look up the corresponding transaction hash.
+pub(crate) fn append_matching_block_logs(
     all_logs: &mut Vec<Log>,
+    provider: impl BlockReader,
     filter: &FilteredParams,
-    block: BlockNumHash,
-    tx_and_receipts: I,
+    block_num_hash: BlockNumHash,
+    receipts: &[Receipt],
     removed: bool,
-) where
-    I: IntoIterator<Item = (TxHash, &'a Receipt)>,
-{
-    let block_number_u256 = U256::from(block.number);
-    // tracks the index of a log in the entire block
+) -> Result<(), FilterError> {
+    // Tracks the index of a log in the entire block.
     let mut log_index: u32 = 0;
-    for (transaction_idx, (transaction_hash, receipt)) in tx_and_receipts.into_iter().enumerate() {
-        for log in receipt.logs.iter() {
-            if log_matches_filter(block, log, filter) {
+
+    // Lazy loaded number of the first transaction in the block.
+    let mut loaded_first_tx_num = None;
+
+    // Iterate over receipts and append matching logs.
+    for (receipt_idx, receipt) in receipts.iter().enumerate() {
+        let logs = &receipt.logs;
+        for log in logs {
+            if log_matches_filter(block_num_hash, log, filter) {
+                let first_tx_num = match loaded_first_tx_num {
+                    Some(num) => num,
+                    None => {
+                        let block_body_indices =
+                            provider.block_body_indices(block_num_hash.number)?.ok_or(
+                                ProviderError::BlockBodyIndicesNotFound(block_num_hash.number),
+                            )?;
+                        loaded_first_tx_num = Some(block_body_indices.first_tx_num);
+                        block_body_indices.first_tx_num
+                    }
+                };
+
+                let transaction_id = first_tx_num + receipt_idx as u64;
```

Comment:
```suggestion
                // This is safe because Transactions and Receipts have the same keys.
                let transaction_id = first_tx_num + receipt_idx as u64;
```